man/masqmail.8
.TH masqmail 8 2015-02-07 masqmail-0.3.5 "Maintenance Commands"

.SH NAME
masqmail \- An offline Mail Transfer Agent

.SH SYNOPSIS

.BR masqmail\  [ \-t ] \  [ \-oi ] \  [ \-f\ \fIADDRESS ]\  \fIRECIPIENT...

.B mailq
.br
.B masqmail \-bp

.B runq
.br
.B masqmail \-q

.BR masqmail\ \-qo\  \fINAME

.BI mailrm\  MSGID...
.br
.BI masqmail\ -Mrm\  MSGID...

.BR masqmail\  [ \-C\ \fIFILE ] \  [ \-odq ]\  \-bs

.BR masqmail\  [ \-C\ \fIFILE ] \  [ \-odq ]\  \-bd\ \-q\fIINTERVAL

(This list is a selection.)


.SH DESCRIPTION

Masqmail is a mail server designed for hosts that do not have a permanent internet connection
e.g. a home network or a single host at home.
It has special support for connections to different ISPs.
It replaces sendmail or other MTAs such as qmail or exim.


.SH OPERATION MODES

Masqmail operates in one of several exclusive modes.

The daemon mode has two flavors that may be,
and usually are, used in combination:
.TP
.B Daemon (listen)
.B \-bd
or invoked as
.B smtpd

Run as daemon.
Act as SMTP server and accept SMTP connections,
on port 25 if not configured differently.

This is normally used in the startup script at system boot
and together with \fB\-q\fIINTERVAL\fR (see below).

Alternatively masqmail can be run with the \fB\-bs\fR option
from inetd.

.TP
.B Daemon (queue)
.B \-q\fIINTERVAL

Run as daemon.
Do regular queue runs at the specified time interval.
This is normally used together with \fB\-bd\fR (see above).

An argument may be a time interval i.e. a numerical value followed by one
of the letters s,m,h,d,w which are interpreted as
seconds, minutes, hours, days or weeks respectively.
Example: \fB\-q30m\fR.
Combinations like \fB\-q1h30m\fR, which sendmail accepts, are not possible.

Alternatively single queue runs (\fB\-q\fR) can be started from cron.

.P
The queue processing mode has two flavors that may be
used in combination:

.TP
.B Single queue run (normal)
.B \-q
(without argument)
or invoked as
.B runq

Do a single queue run.
Try to deliver all messages in the queue.
Masqmail sends to addresses on the local host
and to remote ones, if available online routes permit it.
That means, masqmail sends all queued mail it can.
Online detection is done with the configured
method (see \fBonline_query\fR in \fBmasqmail.conf(5)\fR).

.TP
.B Single queue run (only named route)
.BI \-qo\  NAME

Do a single queue run and deliver only using the specified online route.

The route configuration for \fINAME\fP is read and queued mail
to matching remote recipients is sent.
(See \fBquery_routes.\fINAME\fR in the main config file.)

You may want to use this option in scripts that run as soon as a link
to the internet has been set up (e.g. ip-up).

.B Obsolete behavior:
Up to now: If
.B \-qo
has no argument, the online connection is determinded
by online detect and, if available, mail is sent through it.
If none is available no mail is delivered.

This behavior is likely to change in future versions
because it is included in \-q.
We could do something more useful instead.

.P
The other modes are simple ones:

.TP
.B Rebuild alias database
.B \-bi
or invoked as
.B newaliases

A no-op for masqmail, just exit.
Old sendmail rebuilds its alias database when invoked with this option.
Masqmail reads directly from its alias file
(see config option `alias_file'),
thus no database needs to be updated.

.TP
.B Accept messages
.BR \-bm ,
also the default mode

Accept a text message on stdin.
This is the default mode of operation.
One will hardly use this switch as it is the default.

The command line options
.BR \-f , \-F , \-i / \-oi , \-t
are only used in this mode.

.TP
.B Print queue
.B \-bp
or invoked as
.B mailq

Show the messages in the queue.

.TP
.B Stand-alone SMTP server
.B \-bs

Act as SMTP server by reading commands from stdin
and writing to stdout.
Some mailers (e.g. pine) use this option as an interface.
It can also be used to call masqmail from inetd.

.TP
.B Print version
.B \-bV
or if called without arguments

Print version information, then exit.

.TP
.B Queue manipulation mode
.B \-Mrm
or invoked as
.B mailrm

Remove given messages from the queue.
Privileged users may remove any message, other users only their own.
The message identifiers are listed when printing the queue
(see \fB\-bp\fR).

.P
The default mode:
.P
When no mode had been specified by either one of the above command line
options or by calling masqmail under a special name,
then the default mode \fB\-bm\fR
(i.e. accept messages on stdin) is entered.
However, if neither address arguments are specified nor
\fB\-t\fR is given, then mail can not be sent,
hence something more useful is done: \fB\-bV\fP is assumed.


.SH OPTIONS

Since masqmail is intended to replace sendmail,
it uses the same command line options,
but not all are implemented.
The \fB\-qo\fP option is additional, and unique to masqmail.

.TP
\fB\-\-\fR

Not a `real' option, it means that all following arguments are to be understood
as arguments and not as options even if they begin with a leading dash `\-'.
Mutt is known to call sendmail with this option.

.TP
\fB\-bd\fR

``Daemon (listen)'' mode. See above.

.TP
\fB\-bi\fR

``Rebuild alias database'' mode. See above.

.TP
.B \-bm

``Accept message'' mode. See above.

.TP
\fB\-bp\fR

``Print queue'' mode. See above.

.TP
\fB\-bs\fR

``Stand-alone SMTP server'' mode. See above.

.TP
\fB\-bV \fR

``Print version'' mode. See above.

.TP
\fB\-B \fIarg\fR

\fIarg\fR is usually 8BITMIME.
Some mailers use this to indicate that the message contains characters > 127.
Masqmail is 8-bit clean and ignores this, so you do not have to recompile elm,
which is very painful ;-).
Note though that this violates some conventions:
masqmail does not convert 8 bit messages to any MIME format if it encounters
a mail server which does not advertise its 8BITMIME capability,
masqmail does not advertise this itself.
This is the same practice as that of exim (but different to sendmail).

.TP
\fB\-C \fIfilename\fR

Use another configuration than \fI/etc/masqmail/masqmail.conf\fR.
Useful for debugging purposes.
If not invoked by a privileged user, masqmail will drop all privileges.

.TP
\fB\-d \fInumber\fR

Set the debug level.
This takes precedence before the value of `debug_level' in the configuration file.
Read the warning in the description of the latter.
Only root may set the debug level.

.TP
\fB\-f [\fIaddress\fB]\fR

Set the return path address to \fIaddress\fR.
Only root, the user mail and anyone in group mail is allowed to do that.

.TP
\fB\-F [\fIstring\fB]\fR

Set the full sender name (in the From: header) to \fIstring\fR.

.TP
\fB\-i\fR

Same as \fB\-oi\fR, see below.
Kept for compatibility.

.TP
\fB\-Mrm \fImsgid...\fR

``Queue manipulation'' mode. See above.

.TP
\fB\-m\fR

``Me too''
This switch is ignored as,
masqmail never excludes the sender from any alias expansions.

\fB\-m\fP is an ancient alias for \fB\-om\fP.
Kept for compatibility.

.TP
\fB\-odb\fR

``Deliver in Background''
Masqmail always does this.
Hence masqmail ignores this switch.

.TP
\fB\-odq\fR

``Do Queueing''
Do not attempt to deliver immediately.
Any messages will be queued until the next queue running process picks them
up and delivers them.
You get the same (but global) effect by setting the do_queue option in
/etc/masqmail/masqmail.conf.

.TP
\fB\-oi\fR

A dot as a single character in a line does not terminate the message.

The same as \fB\-i\fP.

.TP
\fB\-oXXX\fR

Any other switch starting with `\-o' is ignored.
This especially affects \-om, \-oem, \-oee.

.TP
\fB\-q

``Single queue run (normal)'' mode. See above.

.TP
\fB\-q\fIINTERVAL

``Daemon (queue)'' mode. See above.

.TP
\fB\-qo \fIname

``Single queue run (only named route)'' mode. See above.

.TP
\fB\-t\fR

Read recipients from mail headers and add them to the ones specified on the
command line.
(Only To:, Cc:, and Bcc: headers are regarded.)

.B WARNING: The behavior changed with version 0.3.1!

In earlier versions command line argument addresses were ``substracted''
from header addresses.

The old behavior was similar to exim's and smail's
(which are anchesters of masqmail).
The new behavior is similar to the one of current postfix versions,
which add the arguments to the set of header recipients.
(Earlier postfix failed in case of address arguments with \-t.)
Sendmail seems to behave differently, depending on the version.

NEWS
This NEWS file lists changes which may be important for you, especially
in regard to compatiblity in upgrades.

See ChangeLog file for more details and technical changes that don't
affect users directly.

The manual pages describe newly added options. Take them for reference.

0.4.0

- Hello, I'm Ossi, the new maintainer and developer of masqmail. Meillo
  handed masqmail over to me, because he wasn't active anymore.


0.3.5

- Adds a glob-pattern based aliasing mechanism.

- Adds route selection based on the From header.

- Changes the paths for PID files and lock files. Both are configurable.


0.3.4

- Reworked the route concept to be more general and have better
  names: ``online_routes'' is now known as ``query_routes'' and the
  ``local_net_route'' in now named ``permanent_routes''. Removed the idea
  of the localnet because it is only part of the more general concept:
  permanent routes. `local_nets' are now represented by `allowed_recipients'
  in a permanent route. (See: examples/localnet.route) Now, there is no
  more abiguity between `local' and `local net'. Run admin/config-transition
  on your config to learn how to update it.

- Switched the default group for masqmail from `trusted' to `mail'.

- Fixed a problem with psb->remote_host, introduced in version 0.3.2

- Fixed a bug in matching of allowed recipients. Thanks to Juergen
  Daubert for finding and reporting the issue.

- Configure options: Enabled auth support per default. Removed support
  for openssl linking. It had been rarely used and could have caused
  legal problems. We don't miss functionality here, because we already
  have builtin MD5 support.  Removed ident support, which is hardly useful
  for masqmail.

- `local_hosts' now understands glob patterns like `*example.org'.


0.3.3

- Fixed a possible security problem, reported by John Lightsey. Unwanted
  identity changes to the user root could have appeared.

- Fixed local-part only address patterns in route conditions, as already
  described in the man page.


0.3.2

- Fixed an important bug with folded headers! In earlier versions mail
  to many recipients may, in unpredictable cases, only have been
  delivered to the ones on the first line. This is fixed now. Sorry for
  the inconvenience.

- Reworked allowed and denied addrs for routes to be more consistent
  and allow more flexible matching:
  - allowed_mail_locals + allowed_return_paths -> allowed_senders
  - not_allowed_mail_locals + not_allowed_return_paths -> denied_senders
  - allowed_rcpt_domains -> allowed_recipients
  - not_allowed_rcpt_domains -> denied_recipients
  See man page masqmail.route(5).

- Reworked online_detect to the simpler online_query. Only pipe is
  supported now. Use
	online_query="/bin/cat /path/to/file"
  instead of
	online_detect=file
	online_file=/path/to/file
  and
	online_query="/path/to/some/script foo"
  instead of
	online_detect=pipe
	online_pipe="/path/to/some/script foo"
  See man page masqmail.conf(5) and admin/config-transition.

- Removed protocol option from route config because it was somehow
  redundant. Now, if `pipe' is set, the protocol will be pipe, otherwise
  it'll be smtp.

- Now the Received headers are much friendlier to read. The header for
  locally (i.e. non-SMTP) received mail is changed to the format postfix
  uses.

- Several improvements to the man pages.

- Updated autoconf.


0.3.1

- Changed the SMTP greeting behavior to match RFC 2821 (always try EHLO
  first). Now always an EHLO greeting is sent first, no matter what kind of
  greeting text the server had sent. If the EHLO failes, an HELO greeting
  is tried as fall back. This change will fix setups in which it hadn't
  been possible to send mail because the server required AUTH but hadn't
  said ``ESMTP'' in its greeting message. Thanks to inne for pointing me
  on that. This fixes also Debian bug #349211.

- The postmaster address is now matched caseless, as required by RFC.

- -qo (without argument) is now considered obsolete. Its behavior (online
  detect and send over the available route) is included in -q. One may
  simply use -q instead. In the -qo case no local mail is sent, but why

src/masqmail.c
// SPDX-FileCopyrightText: (C) 1999-2001 Oliver Kurth
// SPDX-FileCopyrightText: (C) 2010 markus schnalke <meillo@marmaro.de>
// SPDX-License-Identifier: GPL-2.0-or-later
/*
**  MasqMail
*/

#include "masqmail.h"

#include <pwd.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <sys/stat.h>

/*
**  mutually exclusive modes. Note that there is no 'queue daemon' mode.
**  It, as well as the distinction beween the two (non exclusive) daemon
**  (queue and listen) modes, is handled by flags.
*/
enum mta_mode {
	MODE_NONE = 0,  /* to check if a mode was set */
	MODE_ACCEPT,  /* accept message on stdin (fallback mode) */
	MODE_DAEMON,  /* run as daemon */
	MODE_RUNQUEUE,  /* single queue run, online or offline */
	MODE_SMTP,  /* accept SMTP on stdin */
	MODE_LIST,  /* list queue */
	MODE_MCMD,  /* do queue manipulation */
	MODE_VERSION,  /* show version */
	MODE_BI,  /* fake ;-) */
};
enum mta_mode mta_mode = MODE_NONE;

char *pidfile = NULL;
volatile int sigterm_in_progress = 0;

static void
sigterm_handler(int sig)
{
	if (sigterm_in_progress)
		raise(sig);
	sigterm_in_progress = 1;

	if (pidfile) {
		uid_t uid = geteuid();
		if (seteuid(0) != 0) {
			logwrite(LOG_ALERT, "sigterm_handler: could not set "
					"euid to %d: %s\n",
					0, strerror(errno));
		}
		if (unlink(pidfile) != 0)
			logwrite(LOG_WARNING,
					"could not delete pid file %s: %s\n",
					pidfile, strerror(errno));
		seteuid(uid);  /* we exit anyway after this, just to be sure */
	}

	signal(sig, SIG_DFL);
	raise(sig);
}

/*
**  argv: the original argv
**  argp: number of arg (may get modified!)
**  cp: pointing to the char after the option
**    e.g.  `-d 6'     `-d6'
**             ^          ^
*/
gchar*
get_optarg(char *argv[], gint *argp, char *cp)
{
	if (*cp) {
		/* this kind: -xval */
		return cp;
	}
	cp = argv[*argp+1];
	if (cp && (*cp != '-')) {
		/* this kind: -x val */
		(*argp)++;
		return cp;
	}
	return NULL;
}

/*
** Create any missing directory in pathname `dir'. (Like `mkdir -p'.)
** The code is taken from nmh.
*/
gboolean
makedir_rec(char *dir, int perms)
{
	char path[PATH_MAX];
	char *cp, *c;
	int had_an_error = 0;
	mode_t savedmask;

	c = strncpy(path, dir, sizeof(path));

	savedmask = umask(0);

	while (!had_an_error && (c = strchr(c+1, '/'))) {
		*c = '\0';
		/* Create an outer directory. */
		if (mkdir(path, perms) == -1 && errno != EEXIST) {
			fprintf(stderr, "unable to create `%s': %s\n",
					path, strerror(errno));
			had_an_error = 1;
		}
		*c = '/';
	}

	/*
	** Create the innermost nested subdirectory of the
	** path we're being asked to create.
	*/
	if (!had_an_error && mkdir(dir, perms)==-1 && errno != EEXIST) {
		fprintf(stderr, "unable to create `%s': %s\n",
				dir, strerror(errno));
		had_an_error = 1;
	}
	umask(savedmask);

	return (had_an_error) ? 0 : 1;
}

gboolean
write_pidfile(gchar *name)
{
	FILE *fptr;

	if ((fptr = fopen(name, "wt"))) {
		fprintf(fptr, "%d\n", getpid());
		fclose(fptr);
		pidfile = strdup(name);
		return TRUE;
	}
	logwrite(LOG_WARNING, "could not write pid file: %s\n",
			strerror(errno));
	return FALSE;
}

/* on -bd or if -q has an argument */
static void
mode_daemon(gboolean do_listen, gint queue_interval, char *argv[])
{
	guint pid;

	/* daemon */
	if (!conf.run_as_user) {
		if ((conf.orig_uid != 0) && (conf.orig_uid != conf.mail_uid)) {
			fprintf(stderr, "must be root or %s for daemon.\n",
					DEF_MAIL_USER);
			exit(1);
		}
	}

	/* reparent to init only if init is not already the parent */
	if (getppid() != 1) {
		if ((pid = fork()) > 0) {
			exit(0);
		} else if (pid < 0) {
			logwrite(LOG_ALERT, "could not fork!\n");
			exit(1);
		}
	}

	signal(SIGTERM, sigterm_handler);
	makedir_rec(PID_DIR, 0755);
	write_pidfile(PID_DIR "/masqmail.pid");

	conf.do_verbose = FALSE;

	/*
	**  closing and reopening the log ensures that it is open afterwards
	**  because it is possible that the log is assigned to fd 1 and gets
	**  thus closes by fclose(stdout). Similar for the debugfile.
	*/
	logclose();
	fclose(stdin);
	fclose(stdout);
	fclose(stderr);
	logopen();

	logwrite(LOG_NOTICE, "%s %s daemon starting\n", PACKAGE, VERSION);
	listen_port(do_listen ? conf.listen_addresses : NULL,
			queue_interval, argv);
}

/* -bs or called as smtpd or in.smtpd */
static void
mode_smtp()
{
	/* accept smtp message on stdin */
	/* write responses to stderr. */

	struct sockaddr_in saddr;
	gchar *peername = NULL;
	int dummy = sizeof(saddr);

	conf.do_verbose = FALSE;

	if (!conf.run_as_user) {
		set_euidgid(conf.orig_uid, conf.orig_gid, NULL, NULL);
	}

	DEBUG(5) debugf("accepting smtp message on stdin\n");

	if (getpeername(0, (struct sockaddr *) (&saddr), &dummy) == 0) {
		peername = g_strdup(inet_ntoa(saddr.sin_addr));
	} else if (errno != ENOTSOCK)
		exit(1);

	smtp_in(stdin, stderr, peername, NULL);
}

/* default mode if address args or -t is specified, or called as rmail */
static void
mode_accept(address *return_path, gchar *full_sender_name, guint accept_flags,
		char **addresses, int addr_cnt)
{
	/* accept message on stdin */
	accept_error err;
	message *msg = create_message();
	gint i;
	pid_t pid;

	if (return_path && !is_privileged_user(conf.orig_uid)) {
		fprintf(stderr, "must be root, %s or in group %s for "
				"setting return path.\n",
				DEF_MAIL_USER, DEF_MAIL_GROUP);
		exit(1);
	}

	if (!conf.run_as_user) {
		set_euidgid(conf.orig_uid, conf.orig_gid, NULL, NULL);
	}

	DEBUG(5) debugf("accepting message on stdin\n");

	msg->received_prot = PROT_LOCAL;

	/* warn if -t option and cmdline addr args */
	if (addr_cnt && (accept_flags & ACC_RCPT_FROM_HEAD)) {
		logwrite(LOG_ALERT, "command line address arguments are "
				"now *added*  to the mail header\\\n");
		logwrite(LOG_ALERT, "  recipient addresses (instead of "
				"substracted)  when -t is given.\\\n");
		logwrite(LOG_ALERT, "  this changed with version 0.3.1\n");
	}

	for (i = 0; i < addr_cnt; i++) {
		if (addresses[i][0] == '|') {
			logwrite(LOG_ALERT, "no pipe allowed as recipient "
					"address: %s\n", addresses[i]);
			/* should we better ignore this one addr? */
			exit(1);
		}
		msg->rcpt_list = g_list_append(msg->rcpt_list,
				create_address_qualified(addresses[i],
				TRUE, conf.host_name));
	}

	/* -f option */
	msg->return_path = return_path;

	/* -F option */
	msg->full_sender_name = full_sender_name;

	err = accept_message(stdin, msg, accept_flags);

	switch (err) {
	case AERR_OK:
		/* to continue; all other cases exit */
		break;
	case AERR_EOF:
		fprintf(stderr, "unexpected EOF.\n");
		exit(1);
	case AERR_NORCPT:
		fprintf(stderr, "no recipients.\n");
		exit(1);
	case AERR_SIZE:
		fprintf(stderr, "max message size exceeded.\n");
		exit(1);
	default:
		/* should never happen: */
		fprintf(stderr, "Unknown error (%d)\r\n", err);
		exit(1);
	}

	if (!spool_write(msg, TRUE)) {
		fprintf(stderr, "Could not write spool file\n");
		exit(1);
	}

	/* here the mail is queued and thus in our responsibility */
	logwrite(LOG_NOTICE, "%s <= %s with %s\n", msg->uid,
			addr_string(msg->return_path), prot_names[PROT_LOCAL]);

	if (conf.do_queue) {
		/* we're finished as we only need to queue it */
		return;
	}

	/* deliver at once */
	if ((pid = fork()) < 0) {
		logwrite(LOG_ALERT, "could not fork for delivery, id = %s\n",
				msg->uid);
	} else if (pid == 0) {
		conf.do_verbose = FALSE;
		fclose(stdin);
		fclose(stdout);
		fclose(stderr);
		if (deliver(msg)) {
			exit(0);
		} else {
			/*
			**  TODO: Should we really fail here? Because the
			**  mail is queued already. If we fail the client
			**  might submit it again.  If at-once-delivery
			**  is seen as an additional best-effort service,
			**  then we should still exit successful here.
			*/
			exit(1);
		}
	}
}

/*
**  if -Mrm is given
**
**  currently only the `rm' command is supported
**  until this changes, we don't need any facility for further commands
**  return success if at least one message had been deleted
*/
static int
manipulate_queue(char *cmd, char *id[])
{
	gboolean ok = FALSE;

	if (strcmp(cmd, "rm") != 0) {
		fprintf(stderr, "unknown command %s\n", cmd);
		return FALSE;
	}

	set_euidgid(conf.mail_uid, conf.mail_gid, NULL, NULL);

	/* privileged users may delete any mail */
	if (is_privileged_user(conf.orig_uid)) {
		for (; *id; id++) {
			fprintf(stderr, "id: %s\n", *id);
			if (queue_delete(*id)) {
				ok = TRUE;
			}
		}
		return ok;
	}

	struct passwd *pw = getpwuid(conf.orig_uid);
	if (!pw) {
		fprintf(stderr, "could not find a passwd entry for "
				"uid %d: %s\n",
				conf.orig_uid, strerror(errno));
		return FALSE;
	}

	/* non-privileged users may only delete their own messages */
	for (; *id; id++) {
		message *msg = msg_spool_read(*id);

		fprintf(stderr, "id: %s\n", *id);

		if (!msg->ident) {
			fprintf(stderr, "message %s does not have an ident\n",
					*id);
			continue;
		}
		if (strcmp(pw->pw_name, msg->ident) != 0) {
			fprintf(stderr, "you do not own message id %s\n", *id);
			continue;
		}

		if (msg->received_host || (msg->received_prot != PROT_LOCAL)) {
			fprintf(stderr, "message %s was not received "
					"locally\n", *id);
			continue;
		}

		ok = queue_delete(*id);
	}
	return ok;
}

/* -qo, -q (without argument), or called as runq */
static int
run_queue(gboolean do_runq, gboolean do_runq_online, char *route_name)
{
	int ret;

	/* queue runs */
	set_identity(conf.orig_uid, "queue run");

	if (do_runq) {
		ret = queue_run();
	}

	if (do_runq_online) {
		if (route_name) {
			conf.online_query = g_strdup_printf("/bin/echo %s",
					route_name);
		}
		/*
		**  TODO: change behavior of `-qo without argument'?
		**  Because that behavior is included in -q.
		*/
		ret = queue_run_online();
	}
	return ret;
}

/* -bV or default mode if neither addr arg nor -t */
static void
mode_version(void)
{
	gchar *with_resolver = "";
	gchar *with_auth = "";

#ifdef ENABLE_RESOLVER
	with_resolver = " +resolver";
#endif
#ifdef ENABLE_AUTH
	with_auth = " +auth";
#endif

	printf("%s %s%s%s\n", PACKAGE, VERSION, with_resolver, with_auth);
}

void
set_mode(enum mta_mode mode)
{
	if (mta_mode && mta_mode!=mode) {
		fprintf(stderr, "operation mode was already specified "
				"(%d vs. %d)\n", mta_mode, mode);
		exit(1);
	}

	mta_mode = mode;
	return;
}

int
main(int argc, char *argv[])
{
	gchar *progname;
	char *opt;
	gint arg;

	gboolean do_listen = FALSE;
	gboolean do_runq = FALSE;
	gboolean do_runq_online = FALSE;
	gboolean do_queue = FALSE;
	gint queue_interval = 0;
	gchar *M_cmd = NULL;
	gboolean opt_t = FALSE;
	gboolean opt_i = FALSE;
	gchar *conf_file = CONF_FILE;
	gchar *route_name = NULL;
	gchar *f_address = NULL;
	address *return_path = NULL;  /* may be changed by -f option */
	gchar *full_sender_name = NULL;
	gboolean do_verbose = FALSE;
	gint debug_level = -1;

	/* strip the path part */
	progname = strrchr(argv[0], '/');
	progname = (progname) ? progname+1 : argv[0];

	if (strcmp(progname, "mailq") == 0) {
		mta_mode = MODE_LIST;
	} else if (strcmp(progname, "mailrm") == 0) {
		mta_mode = MODE_MCMD;
		M_cmd = "rm";
	} else if (strcmp(progname, "newaliases") == 0) {
		mta_mode = MODE_BI;
	} else if (strcmp(progname, "rmail") == 0) {
		/*
		**  the `rmail' alias should probably be removed now
		**  that we have the rmail script. But let's keep it
		**  for some while for compatibility. 2010-06-19
		*/
		mta_mode = MODE_ACCEPT;
		opt_i = TRUE;
	} else if (strcmp(progname, "runq") == 0) {
		mta_mode = MODE_RUNQUEUE;
		do_runq = TRUE;
	} else if (strcmp(progname, "smtpd") == 0
	           || strcmp(progname, "in.smtpd") == 0) {
		mta_mode = MODE_SMTP;
	}

	/* parse cmd line */
	for (arg=1; arg<argc && argv[arg][0]=='-'; arg++) {
		opt = argv[arg] + 1;  /* points to the char after the dash */

		if (strcmp(opt, "-") == 0) {
			/* everything after `--' are address arguments */
			arg++;
			break;

		} else if (strcmp(opt, "bm") == 0) {
			set_mode(MODE_ACCEPT);

		} else if (strcmp(opt, "bd") == 0) {
			set_mode(MODE_DAEMON);
			do_listen = TRUE;

		} else if (strcmp(opt, "bi") == 0) {
			set_mode(MODE_BI);

		} else if (strcmp(opt, "bs") == 0) {
			set_mode(MODE_SMTP);

		} else if (strcmp(opt, "bp") == 0) {
			set_mode(MODE_LIST);

		} else if (strcmp(opt, "bV") == 0) {
			set_mode(MODE_VERSION);

		} else if (strncmp(opt, "B", 1) == 0) {
			/* we ignore this and throw the argument away */
			get_optarg(argv, &arg, opt+1);

		} else if (strncmp(opt, "C", 1) == 0) {
			conf_file = get_optarg(argv, &arg, opt+1);
			if (!conf_file) {
				fprintf(stderr, "-C requires filename arg.\n");
				exit(1);
			}

		} else if (strncmp(opt, "d", 1) == 0) {
			if (getuid() != 0) {
				fprintf(stderr, "only root may set the "
						"debug level.\n");
				exit(1);
			}
			char *lvl = get_optarg(argv, &arg, opt+1);
			if (!lvl) {
				fprintf(stderr, "-d requires number arg.\n");
				exit(1);
			}
			debug_level = atoi(lvl);

		} else if (strncmp(opt, "f", 1) == 0) {
			/* set return path */
			gchar *address = get_optarg(argv, &arg, opt+1);
			if (!address) {
				fprintf(stderr, "-f requires address arg.\n");
				exit(1);
			}
			f_address = g_strdup(address);

		} else if (strncmp(opt, "F", 1) == 0) {
			full_sender_name = get_optarg(argv, &arg, opt+1);
			if (!full_sender_name) {
				fprintf(stderr, "-F requires name arg.\n");
				exit(1);
			}

		} else if (strcmp(opt, "i") == 0) {
			opt_i = TRUE;

		} else if (strcmp(opt, "m") == 0) {
			/* ignore -m (me too) switch (see man page) */

		} else if (strcmp(opt, "Mrm") == 0) {
			set_mode(MODE_MCMD);
			M_cmd = "rm";

		} else if (strcmp(opt, "odq") == 0) {
			do_queue = TRUE;

		} else if (strcmp(opt, "oi") == 0) {
			opt_i = TRUE;

		} else if (strncmp(opt, "o", 1) == 0) {
			/* ignore all other -oXXX options */

		} else if (strncmp(opt, "qo", 2) == 0) {
			/* must be before the `q' check */
			set_mode(MODE_RUNQUEUE);
			do_runq_online = TRUE;
			/* can be NULL, then we use online detection method */
			/* TODO: behavior might change if it is NULL */
			route_name = get_optarg(argv, &arg, opt+2);
			if (!route_name) {
				fprintf(stderr, "Please do not use -qo "
						"without argument anymore; "
						"use -q instead.\n");
				fprintf(stderr, "The behavior for -qo without "
						"argument is likely to "
						"change.\n");
			}

		} else if (strncmp(opt, "q", 1) == 0) {
			/* must be after the `qo' check */
			gchar *optarg;

			optarg = get_optarg(argv, &arg, opt+1);
			if (optarg) {
				/* do regular queue runs */
				set_mode(MODE_DAEMON);
				queue_interval = time_interval(optarg);
			} else {
				/* do a single queue run */
				set_mode(MODE_RUNQUEUE);
				do_runq = TRUE;
			}

		} else if (strcmp(opt, "t") == 0) {
			opt_t = TRUE;

		} else if (strcmp(opt, "v") == 0) {
			do_verbose = TRUE;

		} else {
			fprintf(stderr, "unrecognized option `-%s'\n", opt);
			exit(1);
		}
	}

	if (!mta_mode && arg==argc && !opt_t) {
		/*
		**  In this case no rcpts can be found, thus no mail
		**  can be sent, thus masqmail will always fail. We
		**  rather do something better instead. This covers
		**  also the case of calling masqmail without args.
		*/
		mode_version();
		exit(0);
	}

	if (mta_mode == MODE_VERSION) {
		mode_version();
		exit(0);
	}

	if (!mta_mode) {
		mta_mode = MODE_ACCEPT;
	}

	/* initialize random generator */
	srand(time(NULL));
	/* ignore SIGPIPE signal */
	signal(SIGPIPE, SIG_IGN);

	/* close all possibly open file descriptors, except std{in,out,err} */
	{
		int i, max_fd = sysconf(_SC_OPEN_MAX);

		if (max_fd <= 0) {
			max_fd = 64;
		}
		for (i=3; i<max_fd; i++) {
			close(i);
		}
	}

	init_conf();

	/*
	**  if we are not privileged, and the config file was changed we
	**  implicetely set the the run_as_user flag and give up all
	**  privileges.
	**
	**  So it is possible for a user to run his own daemon without
	**  breaking security.
	*/
	if ((strcmp(conf_file, CONF_FILE) != 0) && (conf.orig_uid != 0)) {
		logwrite(LOG_NOTICE, "Changing to run_as_user.\n");
		conf.run_as_user = TRUE;
		set_euidgid(conf.orig_uid, conf.orig_gid, NULL, NULL);
		if (setgid(conf.orig_gid)) {
			logwrite(LOG_ALERT, "could not set gid to %d: %s\n",
					conf.orig_gid, strerror(errno));
			exit(1);
		}
		if (setuid(conf.orig_uid)) {
			logwrite(LOG_ALERT, "could not set uid to %d: %s\n",
					conf.orig_uid, strerror(errno));
			exit(1);
		}
	}

	conf.log_dir = LOG_DIR;
	conf.debug_level = debug_level;  /* for debuggin during read_conf() */
	/* FIXME: fails if we run as user */
	logopen();
	if (!read_conf(conf_file)) {
		logwrite(LOG_ALERT, "SHUTTING DOWN due to problems reading "
				"config\n");
		exit(5);
	}
	logclose();

	if (do_queue) {
		conf.do_queue = TRUE;
	}
	if (do_verbose) {
		conf.do_verbose = TRUE;
	}
	if (debug_level >= 0) {  /* if >= 0, it was given by argument */
		conf.debug_level = debug_level;
	}

	/*
	**  It appears that changing to / ensures that we are never in
	**  a directory which we cannot access. This situation could be
	**  possible after changing identity.
	**  Maybe we should only change to / if we not run as user, to
	**  allow relative paths for log files in test setups for
	**  instance.
	*/
	chdir("/");

	if (!conf.run_as_user) {
		if (setgid(0) != 0) {
			fprintf(stderr, "could not set gid to 0. "
					"Is the setuid bit set? : %s\n",
					strerror(errno));
			exit(1);
		}
		if (setuid(0) != 0) {
			fprintf(stderr, "could not gain root privileges. "
					"Is the setuid bit set? : %s\n",
					strerror(errno));
			exit(1);
		}
	}

	if (conf.run_as_user) {
		logwrite(LOG_NOTICE, "Using spool directory `%s' for "
				"lock files.\n", conf.spool_dir);
		conf.lock_dir = conf.spool_dir;
		makedir_rec(conf.spool_dir, 0755);
		makedir_rec(conf.log_dir, 0755);
	} else {
		makedir_rec(conf.lock_dir, 0775);
		chown(conf.lock_dir, conf.mail_uid, conf.mail_gid);
		makedir_rec(conf.spool_dir, 0755);
		chown(conf.spool_dir, conf.mail_uid, conf.mail_gid);
		makedir_rec(conf.log_dir, 0755);
		chown(conf.log_dir, conf.mail_uid, conf.mail_gid);
	}

	if (!logopen()) {
		fprintf(stderr, "could not open log file\n");
		exit(1);
	}

	DEBUG(1) debugf("----STARTING---- masqmail %s\n", VERSION);

	DEBUG(5) {
		gchar **str = argv;
		debugf("args: \n");
		while (*str) {
			debugf("%s \n", *str);
			str++;
		}
	}
	DEBUG(5) debugf("queue_interval = %d\n", queue_interval);

	if (f_address) {
		return_path = create_address_qualified(f_address, TRUE,
				conf.host_name);
		g_free(f_address);
		if (!return_path) {
			fprintf(stderr, "invalid RFC821 address: %s\n",
					f_address);
			exit(1);
		}
	}

	switch (mta_mode) {
	case MODE_DAEMON:
		mode_daemon(do_listen, queue_interval, argv);
		break;

	case MODE_RUNQUEUE:
		exit(run_queue(do_runq, do_runq_online, route_name) ? 0 : 1);
		break;

	case MODE_SMTP:
		mode_smtp();
		break;

	case MODE_LIST:
		queue_list();
		break;

	case MODE_BI:
		exit(0);
		break;  /* well... */

	case MODE_MCMD:
		exit(manipulate_queue(M_cmd, &argv[arg]) ? 0 : 1);
		break;

	case MODE_ACCEPT:
		{
			guint accept_flags = 0;
			accept_flags |= (opt_t ? ACC_RCPT_FROM_HEAD : 0);
			accept_flags |= (opt_i ?
					ACC_DOT_IGNORE : ACC_NODOT_RELAX);
			mode_accept(return_path, full_sender_name,
					accept_flags, argv + arg, argc - arg);
			exit(0);
		}
		break;

	default:
		fprintf(stderr, "unknown mode: %d\n", mta_mode);
		break;
	}

	logclose();

	exit(0);
}
